File: app/src/main/java/com/example/network_termproject/network/NetData.java
package com.example.network_termproject.network;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Base64;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Arrays;

public class NetData {

    private JSONObject data;

    private NetData(){

    }

    public static class Builder{

        private String name = "no";
        private String userId;
        private String type;
        private String content;
        private String chatRoomId;
        private JSONArray list;

        public Builder(){
            list = new JSONArray();
        }

        public Builder setName(String name) {
            this.name = name;
            return this;
        }

        public Builder setType(String type) {
            this.type = type;
            return this;
        }

        public Builder setContent(String content) {
            this.content = content;
            return this;
        }

        public Builder setList(ArrayList<String> dataList) {
            for (String item : dataList){
                this.list.put(item);
            }
            return this;
        }

        public Builder setRoomId(String id){
            this.chatRoomId = id;
            return this;
        }

        public Builder setUserId(String id){
            this.userId = id;
            return this;
        }

        public NetData build() {
            NetData clientData = new NetData();
            clientData.data = new JSONObject();
            try{
                if (name.equals("no")){
                    name = Client.getInstance().getName();
                }
                clientData.data.put("name", name);
                clientData.data.put("type", type);
                clientData.data.put("content", content);
                clientData.data.put("list", list);
                clientData.data.put("roomId", chatRoomId);
                clientData.data.put("userId", userId);
            }
            catch (JSONException e){
                e.printStackTrace();
            }
            return clientData;
        }


        public NetData parseServerData(String data)  {
            NetData receivedData = new NetData();
            try{
                receivedData.data = new JSONObject(data);
            }
            catch (JSONException e){
                e.printStackTrace();
            }
            return receivedData;
        }

        public NetData buildList(){
            NetData clientData = new NetData();
            clientData.data = new JSONObject();
            try{
                clientData.data.put("name", name);
                clientData.data.put("type", type);
                clientData.data.put("content", list);
            }
            catch (JSONException e){
                e.printStackTrace();
            }
            return clientData;
        }
    }

    public JSONObject getData(){
        return data;
    }

    public String getType(){
        try {
            return (String) data.get("type");
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }

    public String getName(){
        try {
            return (String) data.get("name");
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }

    public String getContent(){
        try {
            return (String) data.get("content");
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }

    public String getUserId(){
        try {
            return (String) data.get("userId");
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }

    public String getRoomId(){
        try {
            return (String) data.get("roomId");
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }

    public ArrayList<String> getList(){
        try {
            JSONArray array = (JSONArray) data.get("list");
            ArrayList<String> arr = new ArrayList<>();
            for (int i=0;i<array.length();i++){
                String test = array.getString(i).replace("[", "").replace("]","");
                arr.add(test);
            }
            return arr;
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }

    public Bitmap getImage(){
        try {
            byte[] imageData = Base64.decode(data.getString("content"), Base64.DEFAULT);
            return BitmapFactory.decodeByteArray(imageData, 0, imageData.length);
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }

    public ByteBuffer parseByteBuffer(){
        return ByteBuffer.wrap(data.toString().getBytes());
    }
}

File: app/src/main/java/com/example/network_termproject/network/NetData.kt
package com.example.network_termproject.network

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Base64
import android.util.Log
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.nio.ByteBuffer
import java.util.*
import kotlin.collections.ArrayList

class NetData {

    lateinit var data: JSONObject

    class Builder {
        private var name = "nope"
        private var userID: String? = null
        private var roomID: String? = null
        private lateinit var type: String
        private var content: String? = null
        private var list: ArrayList<String>? = null

        fun setName(name: String): Builder {
            this.name = name
            return this
        }

        fun setType(type: String): Builder {
            this.type = type
            return this
        }

        fun setUserId(id: String): Builder {
            this.userID = id
            return this
        }

        fun setRoomId(id: String): Builder {
            this.roomID = id
            return this
        }

        fun setContent(content: String): Builder {
            this.content = content
            return this
        }

        fun setList(list: ArrayList<String>): Builder {
            this.list = list
            return this
        }

        fun build(): NetData {
            var clientData = NetData()
            clientData.data = JSONObject()
            val idList = JSONArray()
            if (Objects.nonNull(list)) {
                for (id in list!!) {
                    idList.put(id)
                }
            }
            try {
                if (name == "nope") {
                    name = Client.instance.name
                }
                clientData.data!!.apply {
                    put("name", name)
                    put("type", type)
                    put("content", content)
                    put("list", idList)
                    put("roomId", roomID)
                    put("userId", userID)
                }
            } catch (e: JSONException) {
                e.printStackTrace()
            }
            list = null
            return clientData
        }

        fun parseServerData(data: String): NetData {
            val receivedData = NetData()
            try {
                receivedData.data = JSONObject(data)
            } catch (e: JSONException) {
                e.printStackTrace()
            }
            return receivedData
        }
    }

    fun getType(): String {
        return data["type"] as String
    }

    fun getName(): String {
        return data["name"] as String
    }

    fun getContent(): String {
        return data["content"] as String
    }

    fun getUserId(): String {
        return data["userId"] as String
    }

    fun getRoomId(): String {
        return data["roomId"] as String
    }

    fun getList():ArrayList<String>{
        val array = data.getJSONArray("list")
        val arr = ArrayList<String>()
        if (array.length() <= 0){
            Log.e("Size is zero", "Error")
        }
        for (id in 0 until array.length()){
            arr.add(array.getString(id).replace("[", "").replace("]",""))
        }
        return arr
    }

    fun getImage():Bitmap{
        val imageData = data["content"] as String
        val imageArr = Base64.decode(imageData, Base64.DEFAULT)
        return BitmapFactory.decodeByteArray(imageArr,0, imageArr.size)
    }

    fun parseByteBuffer():ByteBuffer{
        return ByteBuffer.wrap(data.toString().toByteArray())
    }
}